### ISSUE TITLE
Various issues with IME

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5-typing#95]
We have already a lot of tickets, but I've been playing with IMEs quite a lot when reviewing #887 and I think that there may be new things here:

## 1 (Firefox)

![mar-29-2017 19-34-46](https://cloud.githubusercontent.com/assets/156149/24468179/b94e47e2-14b7-11e7-90d0-a84cdada9dc5.gif)

## 2 (Firefox)

![mar-29-2017 19-32-28](https://cloud.githubusercontent.com/assets/156149/24468182/b9728300-14b7-11e7-92e8-7b0ed00f18d6.gif)

## 3 (Safari and Chrome)

![mar-29-2017 19-30-44](https://cloud.githubusercontent.com/assets/156149/24468180/b9705eea-14b7-11e7-9159-e0ccd440a8e2.gif)

## 4 (Safari)

![mar-29-2017 19-28-56](https://cloud.githubusercontent.com/assets/156149/24468183/b9747df4-14b7-11e7-9dad-2333ff7b40d9.gif)

## 5 (Chrome)

![mar-29-2017 19-25-56](https://cloud.githubusercontent.com/assets/156149/24468181/b971857c-14b7-11e7-9bec-95365f7c6645.gif)

## 6 (Chrome)

![mar-29-2017 19-13-56](https://cloud.githubusercontent.com/assets/156149/24468184/b97a3d66-14b7-11e7-97a2-020530c253d3.gif)

**Comment by Reinmar** [comment]
@f1ames, could you check these scenarios – whether they are already reported?

**Comment by f1ames** [comment]
The 2nd one (_2 (Firefox)_) looks familiar, however I could not find the issue, the closest one is #89, but it might be a different case.
Also the 4th and 5th (_4 (Safari)_ and _5 (Chrome)_) one seems similar to a problem we have with composition in non-styled elements. As far as I remember it was fixed by #86 or another ticket related to selection. But the cause may be different due to additional styling.

Generally, looks like new cases to handle ;)

**Comment by Reinmar** [comment]
Just stumbled upon 6th case again. This is pretty bad :(

**Comment by Mgsy** [comment]
## Safari

### Steps to reproduce
1. Go to https://ckeditor5.github.io/docs/nightly/ckeditor5/latest/examples/builds/classic-editor.html.
2. Change the language to `spanish`.
3. Select some part of the header and the paragraph.
4. Press <kbd>'</kbd> few times.

### Current result
Text disappears and appears again.

### GIF
![bug_cke5](https://user-images.githubusercontent.com/20988892/31077185-dc395734-a77d-11e7-9192-ae94257eb55b.gif)

### Other information
OS: MacOS X
Browser: Safari
Release: 0.11.0

**Comment by f1ames** [comment]
Went through all reported issue to see if they are still reproducible:

### Issues from https://github.com/ckeditor/ckeditor5-typing/issues/95#issue-217954180

1. Works fine (both using keyboard and mouse). Only difference is that on Safari inserted character is selected (when inserted via keyboard) - as far as I remember it is a native behaviour:
![feb-27-2018 11-37-44-safari](https://user-images.githubusercontent.com/1061942/36726684-f08df4b8-1bba-11e8-85fa-920873ad1048.gif)

2. Works as shown in original gif on Chrome, FF and Safari. Additionally on FF after pressing first <kbd>´</kbd> you have to press <kbd>a</kbd> two times to make it being inserted (first keypress doesn't change anything). Similar to #89.

3. Works in Chrome as shown in an original gif and a little different in Safari:
![feb-27-2018 11-46-19-safari](https://user-images.githubusercontent.com/1061942/36726815-55761022-1bbb-11e8-9393-571f352f6de4.gif)

4. Composition on Safari just breaks in this case:
![feb-27-2018 11-51-01-safari](https://user-images.githubusercontent.com/1061942/36726860-77b88002-1bbb-11e8-82a4-77ce9fe59415.gif)

5. Chrome still broken, Safari behaves similar like in previous case (**4.**) having broken composition, but it inserts two characters instead of one.

6. Behaves as shown on original gif in Chrome. Different on Firefox:
![feb-27-2018 11-56-30-firefox](https://user-images.githubusercontent.com/1061942/36727012-022aa120-1bbc-11e8-8a41-cd4b72e96ed7.gif)
and Safari:
![feb-27-2018 12-03-08-safari](https://user-images.githubusercontent.com/1061942/36727044-1f1e6816-1bbc-11e8-83c8-8a6424b44464.gif)

**To sum up from this 6 issues, first one (*1.*) got resolved in the meantime. Rest is still reproducible.**

### Issue: https://github.com/ckeditor/ckeditor5-typing/issues/95#issuecomment-333519559

**Still reproducible**.

On Chrome pressing <kbd>´</kbd> on such selection doesn't insert any character or change the selection. On Safari it behaves as described in the issue.

**Comment by f1ames** [comment]
While doing some debugging, discovered that the **2 (Firefox)** case from https://github.com/ckeditor/ckeditor5-typing/issues/95#issue-217954180 is caused by the fact of [rerendering children](https://github.com/ckeditor/ckeditor5-engine/blob/2462d27dee6029161853c7f0bbae8953049bd768/src/view/renderer.js#L210) during composition.

**Comment by f1ames** [comment]
When it comes to **6 (Chrome)** issue from the description, it seems that the reason for this is native Chrome issue mentioned here - https://github.com/ckeditor/ckeditor5-engine/issues/898#issuecomment-371509154. When removing `inlineFiller` Chrome moves selection by the amount of characters which were removed - that's why composition starts in some different place...

This does not happen in Safari and Firefox so there the cause must be different (the behaviour is also different).

**Comment by f1ames** [comment]
Extracted **6 (Chrome)** to https://github.com/ckeditor/ckeditor5-engine/issues/1349

**Comment by f1ames** [comment]
My suggestion is to extract unresolved issues (3 and 5) to a separate ticket as they both concern composition on the end of the link and also the last one (https://github.com/ckeditor/ckeditor5-typing/issues/95#issuecomment-333519559) to another separate ticket.

This way we could close this one together with https://github.com/ckeditor/ckeditor5-engine/issues/1417. Also maintaining tickets with one issue (or few issues but with similar cause) is easier that handling above ticket where we have 7 cases (some already or soon fixed). WDYT @Reinmar ?

**Comment by f1ames** [comment]
Extracted link related cases to https://github.com/ckeditor/ckeditor5-typing/issues/149 and issue reported by @Mgsy to https://github.com/ckeditor/ckeditor5-typing/issues/150.

This means we can close this issue together with https://github.com/ckeditor/ckeditor5-engine/issues/1417.

**Comment by Reinmar** [comment]
> My suggestion is to extract unresolved issues (3 and 5) to a separate ticket as they both concern composition on the end of the link and also the last one (#95 (comment)) to another separate ticket.
> 
> This way we could close this one together with ckeditor/ckeditor5-engine#1417. Also maintaining tickets with one issue (or few issues but with similar cause) is easier that handling above ticket where we have 7 cases (some already or soon fixed). WDYT @Reinmar ?

I see I'm late to the party :D But yes, it sounds very good.
